### DISCUSSION TITLE
Render to D3D10/D3D11/OpenGL framebuffer for use with OpenVR? (offscreen rendering)

### DISCUSSION THREAD

**Discussion by OptoCloud** [AvaloniaUI/Avalonia#12648] [Q&A]
I need to pass a framebuffer handle that is a render of an entire window/view handle to OpenVR (Valve VR SDK) to render that into a VR surface.

I have read trough some documentation and searched for solution but haven't found any updated example on how to create such a offscreen surface from Avalonia.

I am looking to feed OpenVR a new framebuffer handle to for every render the window does.
[OpenVR Docs about this](https://github.com/ValveSoftware/openvr/wiki/IVROverlay::SetOverlayTexture)

How would I do this?

Thanks!

**Comment by kekekeks** [comment]
For game engine embedding you need to enable private API usage and provide your own ITopLevelImpl with relevant GPU-backed render target surfaces.

**Reply by OptoCloud** [reply]
Is there any examples or documentation on this?
How to configure AvaloniaUI for the specific graphics apis im targeting and how/where to use this ITopLevelImpl?

**Reply by DenisMontes** [reply]
I need enable private api too.

**Reply by timunie** [reply]
https://github.com/AvaloniaUI/Avalonia/wiki/Using-private-apis-in-nuget-packages

**Reply by OptoCloud** [reply]
> https://github.com/AvaloniaUI/Avalonia/wiki/Using-private-apis-in-nuget-packages

I've already tried this, still can't implement interfaces, its members are private/internal

**Reply by kekekeks** [reply]
`<AvaloniaAccessUnstablePrivateApis>true</AvaloniaAccessUnstablePrivateApis>`. The property from the wiki is to bypass the block for `dotnet pack`
